cmd/soroban-rpc/internal/methods/get_events.go

```diff
@@ -0,0 +1,475 @@
+				// Ignore this token
+				event = event[1:]
+			default:
+				panic("invalid segmentFilter")
+			}
+		} else if segmentFilter.scval != nil {
+			// Exact match the scval
+			if len(event) == 0 || !segmentFilter.scval.Equals(event[0]) {
+				return false
+			}
+			event = event[1:]
+		} else {
+			panic("invalid segmentFilter")
+		}
+	}
+	// Check we had no leftovers
+	return len(event) == 0
+}
+
+type SegmentFilter struct {
+	wildcard *string
+	scval    *xdr.ScVal
+}
+
+func (s *SegmentFilter) UnmarshalJSON(p []byte) error {
+	s.wildcard = nil
+	s.scval = nil
+
+	var tmp string
+	if err := json.Unmarshal(p, &tmp); err != nil {
+		return err
+	}
+	if tmp == "*" {
+		s.wildcard = &tmp
+	} else {
+		var out xdr.ScVal
+		if err := xdr.SafeUnmarshalBase64(tmp, &out); err != nil {
+			return err
+		}
+		s.scval = &out
+	}
+	return nil
+}
+
+type PaginationOptions struct {
+	Cursor string `json:"cursor,omitempty"`
+	Limit  uint   `json:"limit,omitempty"`
+}
+
+type EventStore struct {
+	Client TransactionClient
+}
+
+type TransactionClient interface {
+	Transactions(request horizonclient.TransactionRequest) (horizon.TransactionsPage, error)
+}
+
+// TODO: Extract this to a new package 'eventid'
+// Build a lexically order-able id for this event record. This is
+// based on Horizon's db2/history.Effect.ID method.
+type EventID struct {
+	*toid.ID
+	EventOrder int32
+}
+
+// String returns a string representation of this id
+func (id EventID) String() string {
+	return fmt.Sprintf(
+		"%019d-%010d",
+		id.ToInt64(),
+		id.EventOrder+1,
+	)
+}
+
+func (id *EventID) Parse(input string) error {
+	parts := strings.SplitN(input, "-", 2)
+	if len(parts) != 2 {
+		return fmt.Errorf("invalid event id %s", input)
+	}
+
+	// Parse the first part (toid)
+	idInt, err := strconv.ParseInt(parts[0], 10, 64)
+	if err != nil {
+		return errors.Wrapf(err, "invalid event id %s", input)
+	}
+	parsed := toid.Parse(idInt)
+	id.ID = &parsed
+
+	// Parse the second part (event order)
+	eventOrder, err := strconv.ParseInt(parts[1], 10, 64)
+	if err != nil {
+		return errors.Wrapf(err, "invalid event id %s", input)
+	}
+	// Subtract one to go from the id to the
+	id.EventOrder = int32(eventOrder) - 1
+
+	return nil
+}
+
+func (a EventStore) GetEvents(request GetEventsRequest) ([]EventInfo, error) {
+	if err := request.Valid(); err != nil {
+		return nil, err
+	}
+
+	finish := toid.AfterLedger(request.EndLedger)
+
+	var results []EventInfo
+
+	// TODO: Use a more efficient backend here. For now, we stream all ledgers in
+	// the range from horizon, and filter them. This sucks.
+	cursor := EventID{
+		ID:         toid.New(request.StartLedger, int32(0), 0),
+		EventOrder: 0,
+	}
+	if request.Pagination != nil && request.Pagination.Cursor != "" {
+		if err := cursor.Parse(request.Pagination.Cursor); err != nil {
+			return nil, errors.Wrap(err, "invalid pagination cursor")
+		}
+	}
+	err := a.ForEachTransaction(cursor.ID, finish, func(transaction horizon.Transaction) error {
```

Comment:
just for readability, maybe pull up the inlined function to be declared as package private func instead and reference to it here?

Reply:
That's trickier because it captures a bunch of stuff in the lambda (cursor & pagingToken). Instead I refactored out the event -> eventInfo conversion, to make the inner func more readable.